# Capstone Project by Alan Leon

Using the biggest files from the Gutenberg section in the Github

- Open the file (UTF8 encoding)
- Relative frequency of each character
- Sort by frequency
- Show the sorted characters
- Consider from the marked start and ending
- Measure total time from each book

In [2]:
import os
import re
import time

## Counting Dictionary

In [3]:
charCount = {
'': 0,
'`': 0,
'¶': 0,
'Ã': 0,
'X': 0,
'Ö': 0,
'ß': 0,
')': 0,
'ô': 0,
'Ç': 0,
'E': 0,
']': 0,
'¤': 0,
'æ': 0,
'È': 0,
'*': 0,
':': 0,
'û': 0,
'"': 0,
'b': 0,
'«': 0,
'@': 0,
'è': 0,
'\\': 0,
';': 0,
'6': 0,
'ø': 0,
'ñ': 0,
'»': 0,
'î': 0,
'ï': 0,
'Ó': 0,
'Å': 0,
'|': 0,
'Ú': 0,
'Ä': 0,
'U': 0,
'C': 0,
'¿': 0,
'e': 0,
'%': 0,
'<': 0,
'¯': 0,
'J': 0,
'z': 0,
'à': 0,
'Y': 0,
'ª': 0,
'P': 0,
'~': 0,
'3': 0,
'£': 0,
'M': 0,
'À': 0,
'§': 0,
'Ô': 0,
'=': 0,
',': 0,
'F': 0,
'²': 0,
'm': 0,
'V': 0,
'.': 0,
'l': 0,
'R': 0,
'I': 0,
'7': 0,
'ó': 0,
'Q': 0,
'H': 0,
'd': 0,
'G': 0,
'[': 0,
'h': 0,
'ë': 0,
'N': 0,
'A': 0,
'-': 0,
'Ü': 0,
'5': 0,
'O': 0,
'8': 0,
'0': 0,
't': 0,
'K': 0,
' ': 0,
'L': 0,
'ü': 0,
'Í': 0,
'ã': 0,
'+': 0,
'': 0,
'g': 0,
'^': 0,
'õ': 0,
'{': 0,
'Â': 0,
'#': 0,
'Z': 0,
'á': 0,
'D': 0,
'B': 0,
'ê': 0,
'y': 0,
'Á': 0,
'!': 0,
'Ê': 0,
'p': 0,
'/': 0,
'j': 0,
'å': 0,
'ç': 0,
'É': 0,
'n': 0,
'$': 0,
'ù': 0,
'f': 0,
'T': 0,
'i': 0,
'q': 0,
'ú': 0,
'4': 0,
'o': 0,
'c': 0,
'u': 0,
'í': 0,
'ò': 0,
'Ñ': 0,
'(': 0,
'ý': 0,
'S': 0,
'ö': 0,
'9': 0,
'¡': 0,
'é': 0,
'ä': 0,
'â': 0,
"'": 0,
'r': 0,
'¹': 0,
'}': 0,
'a': 0,
'W': 0,
'2': 0,
'?': 0,
'1': 0,
'_': 0,
'k': 0,
'x': 0,
'v': 0,
'°': 0,
'w': 0,
'&': 0,
's': 0,
'ÿ': 0,
'\x9f':0,
'\n': 0
}

In [4]:
def loadSource(path):
    files = list()
    for file in os.listdir(path):

        parts = file.split(".")			#splits file name into array using the "." as a marker
        if len(parts) >= 2:				#if there are no dots, then the file doesn't even have extension
            tmp = ".".join(parts[:-1])	#everything except last element is sent to a temporary variable
            val = parts[-1] 			##Last element of the parts array is saved in val
            if val.lower() == "txt":	#.lower method to avoid using OR. Checks if the last element in the file name is "txt"
                files.append(file)		
    return files

In [5]:
Times = []
starting = False  

In [6]:
for i in loadSource(r'C:\Users\tomat\OneDrive\Documentos\AmbientesCetys\Capstone'):
    with open('C:\\Users\\tomat\\OneDrive\\Documentos\\AmbientesCetys\\Capstone\\' +i, 'r', encoding='utf-8') as file:
        Times.append(i)
        start_time = time.time() 
        for line in file:
            if '*** START OF THIS PROJECT GUTENBERG' in line:
                starting = True
                continue  # Skip the start line itself

            # Stop reading when the end line is found
            if '*** END OF THIS PROJECT GUTENBERG' in line:
                break  # Stop reading completely
                
            if starting:
                lineData = line
                
                for char in lineData:
                    charCount[char] += 1 # Unfiltered version
        
        top = sorted(charCount.items(), key=lambda item: item[1], reverse=True)
        print(i)
        for key, value in top:
            if value == 0:
                break
            print(f"{key} -> {value}")
        
        charCount = charCount.fromkeys(charCount, 0)
        end_time = time.time() 
        time_taken = end_time - start_time
        Times.append(time_taken)
        starting = False


de-10223-8_UTF8.txt
e -> 886761
  -> 866043
n -> 543386
i -> 413765
r -> 410429
t -> 321142
s -> 306515
a -> 292628
d -> 268447
h -> 235036
u -> 212082
l -> 208687
c -> 163138
g -> 151220
o -> 146041

 -> 145286
m -> 122016
, -> 106563
. -> 98271
b -> 96065
f -> 77195
w -> 60127
z -> 60079
k -> 58274
S -> 50271
p -> 43850
v -> 41912
1 -> 36462
ä -> 32518
ü -> 31201
8 -> 28142
) -> 26038
T -> 25706
( -> 24900
B -> 24708
A -> 23902
G -> 20499
K -> 19897
D -> 19428
M -> 18147
ö -> 17688
ß -> 17671
E -> 16463
P -> 14776
F -> 14551
0 -> 14534
H -> 14146
2 -> 13818
L -> 13601
" -> 13552
7 -> 12791
V -> 12538
R -> 12302
5 -> 12213
W -> 12183
- -> 11767
6 -> 11518
3 -> 11016
4 -> 10913
; -> 8959
y -> 8949
I -> 8769
N -> 8558
Z -> 7468
9 -> 6785
U -> 6201
O -> 5656
J -> 5648
C -> 5424
j -> 5212
: -> 2777
x -> 2561
q -> 1504
Ü -> 1304
é -> 1082
Q -> 1015
' -> 1011
Ö -> 784
Ä -> 622
° -> 596
/ -> 414
á -> 381
= -> 344
X -> 256
[ -> 249
] -> 247
+ -> 226
è -> 216
Y -> 189
ó -> 148
^ -> 139
? -> 13

## Sorting the dictionary using lambda

In [ ]:

# Sort the dictionary by values in descending order
top = sorted(charCount.items(), key=lambda item: item[1], reverse=True)

for key, value in top:
    print(f"{key} -> {value}")

In [9]:
for i in range(0,len(Times) // 2):
    print(Times[i * 2] + " took " + f"{Times[i*2 + 1]} seconds" )

de-10223-8_UTF8.txt took 1.2299768924713135 seconds
en-10012-8_UTF8.txt took 0.08879351615905762 seconds
es-11529-8_UTF8.txt took 0.1264045238494873 seconds
fr-10682-8_UTF8.txt took 0.1289210319519043 seconds


# Activity 2

- Remove newLine
- Convert to lowercase
- Remove punctuation
- Convert accent marks

## Filtered charCount dictionary and replacement table

In [16]:
charCount = {
' ': 0,
'e': 0,
'n': 0,
'i': 0,
'r': 0,
'a': 0,
's': 0,
't': 0,
'd': 0,
'u': 0,
'l': 0,
'o': 0,
'h': 0,
'c': 0,
'g': 0,
'm': 0,
'b': 0,
'f': 0,
'p': 0,
'w': 0,
'k': 0,
'v': 0,
'z': 0,
'1': 0,
'8': 0,
'y': 0,
'j': 0,
'ß': 0,
'q': 0,
'0': 0,
'2': 0,
'7': 0,
'5': 0,
'6': 0,
'3': 0,
'4': 0,
'x': 0,
'9': 0,
'ª': 0,
'²': 0,
'¹': 0,
'æ': 0,
'ø': 0,
}

replacements = {
        'á': 'a', 'à': 'a', 'â': 'a', 'ä': 'a', 'ã': 'a', 'å': 'a',
        'é': 'e', 'è': 'e', 'ê': 'e', 'ë': 'e',
        'í': 'i', 'ì': 'i', 'î': 'i', 'ï': 'i',
        'ó': 'o', 'ò': 'o', 'ô': 'o', 'ö': 'o', 'õ': 'o',
        'ú': 'u', 'ù': 'u', 'û': 'u', 'ü': 'u',
        'ñ': 'n', 'ç': 'c', 'ý': 'y', 'ÿ' : 'y'
    }

## Regex Pattern to ignore punctuation marks

In [17]:

markPattern = r'[^\w\s]'

In [18]:
Times = []
starting = False

In [19]:
for i in loadSource(r'C:\Users\tomat\OneDrive\Documentos\AmbientesCetys\Capstone'):
    with open('C:\\Users\\tomat\\OneDrive\\Documentos\\AmbientesCetys\\Capstone\\' +i, 'r', encoding='utf-8') as file:
        Times.append(i)
        start_time = time.time()
        for line in file:
            if '*** START OF THIS PROJECT GUTENBERG' in line:
                starting = True
                continue  # Skip the start line itself

            # Stop reading when the end line is found
            if '*** END OF THIS PROJECT GUTENBERG' in line:
                break  # Stop reading completely
                
            if starting:
            
                lineData = line.strip()
                lineData = lineData.lower()
                for char in lineData:
                    if re.match(markPattern, char) or char == '_':
                        continue
                    
                    try:
                        replace = replacements[char]
                        charCount[replace] += 1
                        
                    except KeyError:
                        charCount[char] += 1
        top = sorted(charCount.items(), key=lambda item: item[1], reverse=True)
        print(i)
        for key, value in top:
            if value == 0:
                break
            print(f"{key} -> {value}")
        
        charCount = charCount.fromkeys(charCount, 0)
        
        end_time = time.time() 
        time_taken = end_time - start_time
        Times.append(time_taken)
        starting = False

de-10223-8_UTF8.txt
e -> 904651
  -> 866043
n -> 552068
r -> 422731
i -> 422622
s -> 356786
a -> 350410
t -> 346848
d -> 287875
u -> 250848
h -> 249182
l -> 222288
g -> 171719
o -> 170374
c -> 168632
m -> 140163
b -> 120773
f -> 91746
k -> 78171
w -> 72310
z -> 67547
p -> 58626
v -> 54450
1 -> 36462
8 -> 28142
ß -> 17671
0 -> 14534
2 -> 13818
7 -> 12791
5 -> 12213
6 -> 11518
3 -> 11016
4 -> 10913
j -> 10860
y -> 9143
9 -> 6785
x -> 2817
q -> 2519
² -> 28
¹ -> 4
ª -> 1
æ -> 1
ø -> 1
en-10012-8_UTF8.txt
  -> 82981
e -> 55122
t -> 36778
a -> 32999
n -> 31284
o -> 30934
i -> 29372
s -> 29143
r -> 26552
h -> 23084
l -> 19026
d -> 16533
f -> 11636
u -> 11234
g -> 10658
c -> 10642
m -> 10104
w -> 8647
y -> 6997
p -> 6956
b -> 6807
v -> 4634
k -> 3026
x -> 732
0 -> 540
q -> 471
z -> 368
j -> 332
1 -> 152
5 -> 93
2 -> 92
3 -> 63
8 -> 60
4 -> 58
7 -> 43
9 -> 33
6 -> 29
es-11529-8_UTF8.txt
  -> 112505
a -> 77798
e -> 75990
o -> 55800
s -> 43857
n -> 40827
r -> 37454
i -> 35475
l -> 32506
d -> 285

In [ ]:
top = sorted(charCount.items(), key=lambda item: item[1], reverse=True)


for key, value in top:
    print(f"{key} -> {value}")

In [15]:
for i in range(0,len(Times) // 2):
    print(Times[i * 2] + " took " + f"{Times[i*2 + 1]} seconds" )

de-10223-8_UTF8.txt took 12.734052896499634 seconds
en-10012-8_UTF8.txt took 0.9523968696594238 seconds
es-11529-8_UTF8.txt took 1.3221197128295898 seconds
fr-10682-8_UTF8.txt took 1.3342487812042236 seconds


# Done!

As we can see, the german document is the heaviest


 # Results:

 
 ## Unfiltered

 
 ### de-10223-8_UTF8.txt

 - 

 - e -> 886761

 -   -> 866043

 - n -> 543386

 - i -> 413765

 - r -> 410429

 - t -> 321142

 - s -> 306515

 - a -> 292628

 - d -> 268447

 - h -> 235036

 - u -> 212082

 - l -> 208687

 - c -> 163138

 - g -> 151220

 - o -> 146041

 - 

 -  -> 145286

 - m -> 122016

 - , -> 106563

 - . -> 98271

 - b -> 96065

 - f -> 77195

 - w -> 60127

 - z -> 60079

 - k -> 58274

 - S -> 50271

 - p -> 43850

 - v -> 41912

 - 1 -> 36462

 - ä -> 32518

 - ü -> 31201

 - 8 -> 28142

 - ) -> 26038

 - T -> 25706

 - ( -> 24900

 - B -> 24708

 - A -> 23902

 - G -> 20499

 - K -> 19897

 - D -> 19428

 - M -> 18147

 - ö -> 17688

 - ß -> 17671

 - E -> 16463

 - P -> 14776

 - F -> 14551

 - 0 -> 14534

 - H -> 14146

 - 2 -> 13818

 - L -> 13601

 - " -> 13552

 - 7 -> 12791

 - V -> 12538

 - R -> 12302

 - 5 -> 12213

 - W -> 12183

 - - -> 11767

 - 6 -> 11518

 - 3 -> 11016

 - 4 -> 10913

 - ; -> 8959

 - y -> 8949

 - I -> 8769

 - N -> 8558

 - Z -> 7468

 - 9 -> 6785

 - U -> 6201

 - O -> 5656

 - J -> 5648

 - C -> 5424

 - j -> 5212

 - : -> 2777

 - x -> 2561

 - q -> 1504

 - Ü -> 1304

 - é -> 1082

 - Q -> 1015

 - ' -> 1011

 - Ö -> 784

 - Ä -> 622

 - ° -> 596

 - / -> 414

 - á -> 381

 - = -> 344

 - X -> 256

 - [ -> 249

 - ] -> 247

 - + -> 226

 - è -> 216

 - Y -> 189

 - ó -> 148

 - ^ -> 139

 - ? -> 138

 - ñ -> 124

 - â -> 123

 - à -> 119

 - Ã -> 94

 - § -> 90

 - | -> 89

 - ç -> 68

 - É -> 55

 - ô -> 47

 - ! -> 47

 - ë -> 46

 - $ -> 46

 - ï -> 41

 - } -> 34

 - ú -> 32

 - * -> 30

 - { -> 29

 - ² -> 28

 - ê -> 28

 - í -> 28

 - û -> 26

 - _ -> 23

 - ¶ -> 22

 - ¤ -> 22

 - î -> 19

 - \ -> 14

 - \# -> 13

 - å -> 11

 - ~ -> 8

 - ò -> 7

 - ã -> 6

 - % -> 4

 - < -> 4

 - ý -> 4

 - ¹ -> 4

 - Á -> 3

 - Ç -> 2

 - õ -> 2

 - Â -> 2

 - ù -> 2

 - ¡ -> 2

 - & -> 2

 - ` -> 1

 - æ -> 1

 - ø -> 1

 - Ó -> 1

 - Å -> 1

 - ¯ -> 1

 - ª -> 1

 - £ -> 1

 - ÿ -> 1

 - Ÿ -> 1

 ### en-10012-8_UTF8.txt

 -   -> 83338

 - e -> 54465

 - t -> 35518

 - a -> 32291

 - n -> 30656

 - o -> 30411

 - s -> 28276

 - i -> 28041

 - r -> 26124

 - h -> 22715

 - l -> 18761

 - d -> 16284

 - f -> 11362

 - u -> 11106

 - g -> 10453

 - c -> 10295

 - m -> 9723

 - 

 -  -> 8763

 - w -> 8400

 - , -> 7730

 - y -> 6850

 - p -> 6635

 - b -> 6474

 - v -> 4512

 - k -> 2954

 - . -> 2791

 - - -> 1760

 - I -> 1331

 - T -> 1260

 - S -> 867

 - x -> 714

 - A -> 708

 - E -> 639

 - 0 -> 540

 - ; -> 531

 - O -> 523

 - N -> 503

 - q -> 454

 - R -> 428

 - M -> 381

 - H -> 369

 - z -> 354

 - C -> 347

 - B -> 333

 - P -> 321

 - F -> 274

 - L -> 265

 - j -> 254

 - D -> 249

 - W -> 247

 - _ -> 206

 - G -> 205

 - " -> 204

 - ' -> 160

 - 1 -> 152

 - Y -> 147

 - U -> 128

 - V -> 122

 - ñ -> 120

 - 5 -> 93

 - 2 -> 92

 - J -> 78

 - : -> 76

 - K -> 72

 - 3 -> 63

 - 8 -> 60

 - ] -> 59

 - [ -> 59

 - 4 -> 58

 - ) -> 46

 - ( -> 46

 - 7 -> 43

 - 9 -> 33

 - ! -> 32

 - 6 -> 29

 - ° -> 22

 - X -> 18

 - Q -> 17

 - Z -> 14

 - é -> 13

 - ? -> 13

 - / -> 8

 - $ -> 6

 - ë -> 5

 - Ñ -> 5

 ### es-11529-8_UTF8.txt

 -   -> 112689

 - a -> 74081

 - e -> 72312

 - o -> 50755

 - s -> 42938

 - n -> 38992

 - r -> 36901

 - l -> 31520

 - i -> 30588

 - d -> 28045

 - u -> 24364

 - t -> 23933

 - c -> 22231

 - m -> 16384

 - 

 -  -> 14649

 - p -> 13274

 - b -> 9819

 - . -> 9644

 - , -> 7993

 - q -> 6573

 - g -> 6216

 - h -> 5760

 - v -> 5712

 - - -> 5051

 - ó -> 4809

 - í -> 4742

 - y -> 4581

 - f -> 3574

 - j -> 3225

 - z -> 2595

 - á -> 2548

 - é -> 2211

 - E -> 1465

 - C -> 1317

 - ñ -> 1292

 - A -> 1166

 - P -> 1157

 - L -> 986

 - S -> 919

 - x -> 753

 - ! -> 647

 - ¡ -> 646

 - ¿ -> 637

 - ? -> 637

 - ú -> 575

 - M -> 574

 - R -> 553

 - N -> 543

 - : -> 537

 - ; -> 528

 - D -> 468

 - Y -> 420

 - T -> 357

 - _ -> 334

 - H -> 291

 - V -> 252

 - O -> 234

 - Q -> 227

 - F -> 190

 - " -> 181

 - U -> 174

 - I -> 141

 - B -> 137

 - G -> 102

 - J -> 92

 - ) -> 57

 - ( -> 57

 - ª -> 50

 - * -> 45

 - ü -> 40

 - \# -> 32

 - X -> 21

 - k -> 10

 - Z -> 6

 - 8 -> 3

 - ' -> 3

 - 2 -> 3

 - 1 -> 3

 - ô -> 2

 - « -> 2

 - » -> 2

 - î -> 2

 - Í -> 2

 - É -> 2

 - â -> 2

 - 6 -> 1

 - Ú -> 1

 - 3 -> 1

 - 0 -> 1

 - K -> 1

 - Á -> 1

 - 4 -> 1

 - 9 -> 1

 - w -> 1

 ### fr-10682-8_UTF8.txt

 -   -> 112656

 - e -> 85067

 - s -> 51485

 - a -> 46315

 - n -> 42171

 - t -> 40362

 - i -> 38476

 - r -> 38126

 - u -> 37722

 - o -> 30974

 - l -> 29517

 - d -> 19931

 - c -> 17289

 - m -> 15864

 - p -> 15833

 - 

 -  -> 14010

 - é -> 11045

 - v -> 10804

 - , -> 8897

 - ' -> 7777

 - . -> 7358

 - q -> 5621

 - f -> 5583

 - g -> 5520

 - b -> 5311

 - h -> 5189

 - - -> 4970

 - j -> 2964

 - à -> 2803

 - x -> 2517

 - è -> 2094

 - y -> 1666

 - ! -> 1538

 - ; -> 1438

 - L -> 1304

 - ê -> 1171

 - I -> 1163

 - _ -> 1049

 - z -> 1029

 - C -> 916

 - J -> 888

 - N -> 848

 - S -> 843

 - E -> 761

 - A -> 709

 - ? -> 694

 - M -> 602

 - D -> 560

 - P -> 529

 - ç -> 467

 - O -> 415

 - ô -> 413

 - V -> 370

 - â -> 370

 - î -> 355

 - û -> 352

 - : -> 314

 - R -> 312

 - U -> 309

 - Q -> 267

 - ù -> 242

 - T -> 236

 - G -> 174

 - B -> 148

 - H -> 124

 - w -> 118

 - k -> 108

 - X -> 78

 - F -> 71

 - ñ -> 56

 - ) -> 51

 - ( -> 51

 - * -> 50

 - É -> 47

 - Z -> 46

 - 1 -> 42

 - ] -> 37

 - [ -> 37

 - « -> 35

 - » -> 32

 - Ç -> 31

 - ï -> 27

 - Y -> 19

 - W -> 18

 - 0 -> 14

 - K -> 14

 - Ê -> 8

 - È -> 6

 - À -> 3

 - ë -> 3

 - 6 -> 2

 - 5 -> 2

 - 3 -> 1

 - Ô -> 1

 - 8 -> 1

 - 9 -> 1

 
 ## Filtered

 
 ### de-10223-8_UTF8.txt

 - e -> 904651

 -   -> 866043

 - n -> 552068

 - r -> 422731

 - i -> 422622

 - s -> 356786

 - a -> 350410

 - t -> 346848

 - d -> 287875

 - u -> 250848

 - h -> 249182

 - l -> 222288

 - g -> 171719

 - o -> 170374

 - c -> 168632

 - m -> 140163

 - b -> 120773

 - f -> 91746

 - k -> 78171

 - w -> 72310

 - z -> 67547

 - p -> 58626

 - v -> 54450

 - 1 -> 36462

 - 8 -> 28142

 - ß -> 17671

 - 0 -> 14534

 - 2 -> 13818

 - 7 -> 12791

 - 5 -> 12213

 - 6 -> 11518

 - 3 -> 11016

 - 4 -> 10913

 - j -> 10860

 - y -> 9143

 - 9 -> 6785

 - x -> 2817

 - q -> 2519

 - ² -> 28

 - ¹ -> 4

 - ª -> 1

 - æ -> 1

 - ø -> 1

 - 

 ### en-10012-8_UTF8.txt

 -   -> 82981

 - e -> 55122

 - t -> 36778

 - a -> 32999

 - n -> 31284

 - o -> 30934

 - i -> 29372

 - s -> 29143

 - r -> 26552

 - h -> 23084

 - l -> 19026

 - d -> 16533

 - f -> 11636

 - u -> 11234

 - g -> 10658

 - c -> 10642

 - m -> 10104

 - w -> 8647

 - y -> 6997

 - p -> 6956

 - b -> 6807

 - v -> 4634

 - k -> 3026

 - x -> 732

 - 0 -> 540

 - q -> 471

 - z -> 368

 - j -> 332

 - 1 -> 152

 - 5 -> 93

 - 2 -> 92

 - 3 -> 63

 - 8 -> 60

 - 4 -> 58

 - 7 -> 43

 - 9 -> 33

 - 6 -> 29

 - 

 ### es-11529-8_UTF8.txt

 -   -> 112505

 - a -> 77798

 - e -> 75990

 - o -> 55800

 - s -> 43857

 - n -> 40827

 - r -> 37454

 - i -> 35475

 - l -> 32506

 - d -> 28513

 - u -> 25154

 - t -> 24290

 - c -> 23548

 - m -> 16958

 - p -> 14431

 - b -> 9956

 - q -> 6800

 - g -> 6318

 - h -> 6051

 - v -> 5964

 - y -> 5001

 - f -> 3764

 - j -> 3317

 - z -> 2601

 - x -> 774

 - ª -> 50

 - k -> 11

 - 1 -> 3

 - 8 -> 3

 - 2 -> 3

 - w -> 1

 - 0 -> 1

 - 6 -> 1

 - 3 -> 1

 - 4 -> 1

 - 9 -> 1

 - 

 ### fr-10682-8_UTF8.txt

 -   -> 112584

 - e -> 100202

 - s -> 52328

 - a -> 50200

 - n -> 43075

 - t -> 40598

 - i -> 40021

 - u -> 38625

 - r -> 38438

 - o -> 31803

 - l -> 30821

 - d -> 20491

 - c -> 18703

 - m -> 16466

 - p -> 16362

 - v -> 11174

 - q -> 5888

 - g -> 5694

 - f -> 5654

 - b -> 5459

 - h -> 5313

 - j -> 3852

 - x -> 2595

 - y -> 1685

 - z -> 1075

 - w -> 136

 - k -> 122

 - 1 -> 42

 - 0 -> 14

 - 5 -> 2

 - 6 -> 2

 - 8 -> 1

 - 3 -> 1

 - 9 -> 1

 - 